api/solarperformanceinsight_api/utils.py

```diff
@@ -0,0 +1,197 @@
+        try:
+            float(header)
+        except ValueError:
+            pass
+        else:
+            raise HTTPException(
+                status_code=400,
+                detail=(
+                    f"The header '{header}' can be parsed as a "
+                    "float indicating a header row may be missing?"
+                ),
+            )
+        if len(header) == 0 or header.startswith("Unnamed:"):
+            raise HTTPException(status_code=400, detail=f"Empty header for column {i}")
+
+    if "time" in head_df.columns:
+        kwargs["parse_dates"] = ["time"]
+    content.seek(0)
+    try:
+        df = pd.read_csv(content, **kwargs)  # type: ignore
+    except (EmptyDataError, ParserError) as err:
+        raise HTTPException(status_code=400, detail=err.args[0])
+    if df.empty:
+        raise HTTPException(status_code=400, detail="Empty CSV file")
+    return df
+
+
+def read_arrow(content: IO) -> pd.DataFrame:
+    """Read a buffer in Apache Arrow File format into a DataFrame"""
+    try:
+        table = pa.ipc.open_file(content).read_all()
+    except pa.lib.ArrowInvalid as err:
+        raise HTTPException(status_code=400, detail=err.args[0])
+    df = table.to_pandas(split_blocks=True)
+    return df
+
+
+def verify_content_type(content_type: str) -> Callable[[IO], pd.DataFrame]:
+    """Checks if we can read the content_type and returns the appropriate function for
+    reading"""
+    csv_types = ("text/csv", "application/vnd.ms-excel")
+    arrow_types = (
+        "application/octet-stream",
+        "application/vnd.apache.arrow.file",
+    )
+    # reject non csv/arrow
+    if content_type not in csv_types and content_type not in arrow_types:
+        raise HTTPException(
+            status_code=415,
+            detail=(
+                "Acceptable formats are CSV (text/csv) and the Apache Arrow "
+                "file format (application/vnd.apache.arrow.file)"
+            ),
+        )
+    if content_type in csv_types:
+        return read_csv
+    else:
+        return read_arrow
+
+
+def validate_dataframe(df: pd.DataFrame, columns: List[str]) -> Set[str]:
+    """Validates that the input dataframe has all given columns, that the
+    'time' column has a datetime type and no microseconds, and that all
+    other columns are floats"""
+    expected = set(columns)
+    actual = set(df.columns)
+    diff = expected - actual
+    if len(diff) > 0:
+        raise HTTPException(
+            status_code=400, detail="Data is missing column(s) " + ", ".join(diff)
+        )
+    if "time" in expected:
+        if not is_datetime64_any_dtype(df["time"]):
+            raise HTTPException(
+                status_code=400,
+                detail='"time" column could not be parsed as a timestamp',
+            )
+        if not (df["time"].dt.microsecond == 0).all():  # type: ignore
+            raise HTTPException(
+                status_code=400, detail='"time" column must only have second precision'
```

Comment:
I could see this being a warning that time is being rounded for the user, instead of an Exception.

Reply:
Yeah, this I doubt many users would upload microseconds in the first place. I removed this check and just round them off when reindexing